crates/ruff_python_index/src/fstring_ranges.rs

```diff
@@ -0,0 +1,61 @@
+use std::collections::BTreeMap;
+
+use ruff_python_parser::Tok;
+use ruff_text_size::{TextRange, TextSize};
+
+/// Stores the ranges of all f-strings in a file sorted by [`TextRange::start`].
+/// There can be multiple overlapping ranges for nested f-strings.
+#[derive(Debug)]
+pub struct FStringRanges {
+    raw: BTreeMap<TextSize, TextRange>,
+}
+
+impl FStringRanges {
+    /// Return the [`TextRange`] of the innermost f-string at the given offset.
+    pub fn innermost(&self, offset: TextSize) -> Option<TextRange> {
+        self.raw
+            .range(..=offset)
+            .rev()
+            .find(|(_, range)| range.contains(offset))
+            .map(|(_, range)| *range)
+    }
+
+    /// Return the [`TextRange`] of the outermost f-string at the given offset.
+    pub fn outermost(&self, offset: TextSize) -> Option<TextRange> {
+        self.raw
+            .range(..=offset)
+            .find(|(_, range)| range.contains(offset))
+            .map(|(_, range)| *range)
```

Comment:
This has a worst time complexity of `O(n)` if the searched offset is the last in the map. 

I think it would be better to iterate from the back and take the last entry where `range.contains` is still true. This should be `O(log(n))` for almost all entries

Reply:
Yes, this is correct. Thanks for the suggestion. One thing to note is that we'll have to skip the initial ranges where the offset doesn't belong because those will be part of the inner ranges while the offset might be part of an outer range. The iterator of nested ranges goes from the innermost to the outermost.

For example:
```python
#                                                     v
  f"normal" f"another" f"first {f"second {f"third"} second"} first"
#                                         ^^^1^^^^
#                               ^^^^^^^^^^^^^2^^^^^^^^^^^^^
#                      ^^^^^^^^^^^^^^^^^^^^^^3^^^^^^^^^^^^^^^^^^^^^
#           ^^^^4^^^^^
# ^^^^5^^^^
```

If the offset is where `v` is at, then the iterator would yield the range in the order mentioned above. Now, the offset doesn't belong in (1) so we will skip until we get a range where the offset belongs to. Then, we'll continue until we get a range which _doesn't_ contain the offset and the last range will be the outermost range.

So, skip (1), next (2), next (3), stop. Outermost range is (3).